sql: use new explain format in explain bundles

RaduBerinde opened the pull request:
#### util: treeprinter improvements and new style

This change cleans up the treeprinter implementation and interface and
adds a "with bullets" style that looks like this:
```
  • foo
  │
  ├── • bar1
  │   │ bar2
  │   │
  │   └── • baz
  │
  └── • qux
```
This is a preliminary design from Anne for the new EXPLAIN output.

Release note: None

#### opt: rework explain output test

Changing the explain output test from an Example to a datadriven test
(which is much easier to update).

Release note: None

#### sql: use new explain format in explain bundles

This change switches the plan string in EXPLAIN ANALYZE (DEBUG)
bundles to use the new style with bullets. See the updated test for
some examples.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55637)
<!-- Reviewable:end -->

asubiotto reviewed (changes requested):
Sorry for the delay in reviewing this.

> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-:-MKFP1bt7p4_db-wWHfZ:b8a4juy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/exec/explain/output_test.go, line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-MKFSRBz0jhSvvNXlsRl:-MKFSRBz0jhSvvNXlsRm:by9p9vl) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbe0e32db1558600c25640c53c8d10bce89ea195/pkg/sql/opt/exec/explain/output_test.go#L30)):*
> ```Go
> 
> func TestOutputBuilder(t *testing.T) {
> 	example := func(flags explain.Flags) *explain.OutputBuilder {
> ```

Seems like this should be an argument to the datadriven test, but you have more context on the utility of that. Consider leaving a TODO.
___
*[pkg/util/treeprinter/tree_printer.go, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-MKFPPYFDDKFmeQu6QBR:-MKFPPYFDDKFmeQu6QBS:b-ne0cel) ([raw file](https://github.com/cockroachdb/cockroach/blob/89cb106cbdac0e9fd161c100e011ea2440b98afe/pkg/util/treeprinter/tree_printer.go#L61)):*
> ```Go
> 
> // Style is one of the predefined treeprinter styles.
> type Style struct {
> ```

Why use a struct instead of:
```
type Style int
const (
    DefaultStyle Style = iota
    CompactStyle
    BulletStyle
)
```
Seems like the styles are mutually exclusive.
___
*[pkg/util/treeprinter/tree_printer.go, line 141 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-MKFPv-S7qsEf43pzPqY:-MKFPv-S7qsEf43pzPqZ:bem31xb) ([raw file](https://github.com/cockroachdb/cockroach/blob/89cb106cbdac0e9fd161c100e011ea2440b98afe/pkg/util/treeprinter/tree_printer.go#L141)):*
> ```Go
> 	rows [][]rune
> 
> 	// Information pertaining to the last node at a given depth.
> ```

OOC, what does this comment mean by `last node`?
___
*[pkg/util/treeprinter/tree_printer.go, line 152 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-MKFRKEk-HaJ1mpAuqhv:-MKFRKEl1kWcCB2kC-TR:birocl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/89cb106cbdac0e9fd161c100e011ea2440b98afe/pkg/util/treeprinter/tree_printer.go#L152)):*
> ```Go
> 	// firstChildConnectRow is the index (in tree.rows) of the row up to which we
> 	// have to connect the first child of this node.
> 	firstChildConnectRow int
> ```

I don't know if I'm missing this, but it seems like there isn't a "high-level" comment about this treeprinter. My opinion is that this file could benefit from such a comment (and maybe walk through an example), but I don't think it's a requirement, so I'll leave that up to you.
___
*[pkg/util/treeprinter/tree_printer.go, line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-MKFQ6yiDa_f7SYpqk5Z:-MKFQ6yiDa_f7SYpqk5_:b-hi46lh) ([raw file](https://github.com/cockroachdb/cockroach/blob/89cb106cbdac0e9fd161c100e011ea2440b98afe/pkg/util/treeprinter/tree_printer.go#L161)):*
> ```Go
> 	//      foo
> 	//       └── bar1               <---- nextSiblingConnectRow
> 	//           bar2               <---- firstChildConnectRow
> ```

Something seems to be up with this `bar2` indentation throughout the file, or is that normal?


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: use new explain format in explain bundles

This change switches the plan string in EXPLAIN ANALYZE (DEBUG)
bundles to use the new style with bullets. See the updated test for
some examples.

Release note: None

RaduBerinde committed:
opt: rework explain output test

Changing the explain output test from an Example to a datadriven test
(which is much easier to update).

Release note: None

RaduBerinde committed:
util: treeprinter improvements and new style

This change cleans up the treeprinter implementation and interface and
adds a "with bullets" style that looks like this:
```
  • foo
  │
  ├── • bar1
  │   │ bar2
  │   │
  │   └── • baz
  │
  └── • qux
```
This is a preliminary design from Anne for the new EXPLAIN output.

Release note: None

RaduBerinde reviewed (commented):
Thanks for reviewing this tediousness :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-:-MKFe-9y1cFxQWSOtqs4:b-sodrp8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @mgartner)

___
*[pkg/sql/opt/exec/explain/output_test.go, line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-MKFSRBz0jhSvvNXlsRl:-MKFYfFC07A1zrK6CqNh:b-x7z8jn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbe0e32db1558600c25640c53c8d10bce89ea195/pkg/sql/opt/exec/explain/output_test.go#L30)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Seems like this should be an argument to the datadriven test, but you have more context on the utility of that. Consider leaving a TODO.
</blockquote></details>

Not sure I'm understanding correctly - an argument in the test file? That would be nice but I'd have to invent and parse a language that specifies the OutputBuilder calls, no? I really only want the `-rewrite` functionality, not so much the ability to write arbitrary tests.
___
*[pkg/util/treeprinter/tree_printer.go, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-MKFPPYFDDKFmeQu6QBR:-MKF_EJf9qy-kUik2-xp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/89cb106cbdac0e9fd161c100e011ea2440b98afe/pkg/util/treeprinter/tree_printer.go#L61)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Why use a struct instead of:
```
type Style int
const (
    DefaultStyle Style = iota
    CompactStyle
    BulletStyle
)
```
Seems like the styles are mutually exclusive.
</blockquote></details>

Done.
___
*[pkg/util/treeprinter/tree_printer.go, line 141 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-MKFPv-S7qsEf43pzPqY:-MKFcK1qBcuKhuVkai9G:b8jtgid) ([raw file](https://github.com/cockroachdb/cockroach/blob/89cb106cbdac0e9fd161c100e011ea2440b98afe/pkg/util/treeprinter/tree_printer.go#L141)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

OOC, what does this comment mean by `last node`?
</blockquote></details>

Updated, and hopefully the new bigger comment helps.
___
*[pkg/util/treeprinter/tree_printer.go, line 152 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-MKFRKEk-HaJ1mpAuqhv:-MKFcN5-C48Wp_Q2VoLh:b-wes99) ([raw file](https://github.com/cockroachdb/cockroach/blob/89cb106cbdac0e9fd161c100e011ea2440b98afe/pkg/util/treeprinter/tree_printer.go#L152)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't know if I'm missing this, but it seems like there isn't a "high-level" comment about this treeprinter. My opinion is that this file could benefit from such a comment (and maybe walk through an example), but I don't think it's a requirement, so I'll leave that up to you.
</blockquote></details>

You're right, I added a high level comment for `tree` which describes the high level mechanics of what's going on.
___
*[pkg/util/treeprinter/tree_printer.go, line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-MKFQ6yiDa_f7SYpqk5Z:-MKF_3-KBw26AIVOjMK1:bvhammj) ([raw file](https://github.com/cockroachdb/cockroach/blob/89cb106cbdac0e9fd161c100e011ea2440b98afe/pkg/util/treeprinter/tree_printer.go#L161)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Something seems to be up with this `bar2` indentation throughout the file, or is that normal?
</blockquote></details>

Looks fine in my terminal and in github, probably some issue with reviewable.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 6 of 6 files at r5, 2 of 2 files at r6, 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-:-MKFsT0i7qs2PBk2TSnH:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/util/treeprinter/tree_printer.go, line 141 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-MKFPv-S7qsEf43pzPqY:-MKFsGjN6SEjDvxQx_1X:b-ugqjzl) ([raw file](https://github.com/cockroachdb/cockroach/blob/89cb106cbdac0e9fd161c100e011ea2440b98afe/pkg/util/treeprinter/tree_printer.go#L141)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Updated, and hopefully the new bigger comment helps.
</blockquote></details>

Awesome, looks great


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55637#-:-MKG7GLiC8Hp87Qom9EB:b-814g8k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @mgartner)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
